src/llm/apis/openai_completions.cpp

```diff
@@ -278,6 +278,77 @@ absl::Status OpenAIChatCompletionsHandler::parseMessages() {
     return absl::OkStatus();
 }
 
+absl::Status OpenAIChatCompletionsHandler::parseTools() {
+    auto it = doc.FindMember("tools");
+    if (it == doc.MemberEnd())
+        return absl::OkStatus();
+    if (!it->value.IsArray())
+        return absl::InvalidArgumentError("Tools are not an array");
+    auto tool_choice_it = doc.FindMember("tool_choice");
+    std::string tool_choice{"auto"};
+    if (tool_choice_it != doc.MemberEnd()) {
+        if (tool_choice_it->value.IsString()){
+            tool_choice = tool_choice_it->value.GetString();
+            if (tool_choice != "none" && tool_choice != "auto")
+                return absl::InvalidArgumentError("tool_choice should be either 'none' or 'auto'");
+        } else if (tool_choice_it->value.IsObject()) {
+            auto tool_choice_functionIt = tool_choice_it->value.GetObject().FindMember("function");
+            if (tool_choice_functionIt != tool_choice_it->value.GetObject().MemberEnd() && tool_choice_functionIt->value.IsObject()) {
+                auto nameIt = tool_choice_functionIt->value.GetObject().FindMember("name");
+                if (nameIt != tool_choice_functionIt->value.GetObject().MemberEnd() && nameIt->value.IsString()) {
+                    tool_choice = nameIt->value.GetString();
+                } else {
+                    return absl::InvalidArgumentError("tool_choice.function.name is not a valid string");
+                }
+            } else {
+                return absl::InvalidArgumentError("tool_choice.function is not a valid JSON object");
+            }
+        } else
+            return absl::InvalidArgumentError("tool_choice is not a valid JSON object or string");
+    }
+    std::cout << "Tool choice: " << tool_choice << std::endl;
+
+    bool jsonChanged = false;
+    if (tool_choice == "auto")  // for now, with auto choice we don't need to do anything
+        return absl::OkStatus();
+    if (tool_choice == "none") {
+        // remove tools from the request
+        doc.RemoveMember("tools");
```

Comment:
Is that call safe? We remove `tools` from the docs, but don't we have invalid `it` after that operation?
Shouldn't we exit here and not get into below loop?